### ISSUE TITLE
hyperkit fails to build from source: sandbox_apply_container operation not permitted

### ISSUE THREAD

**Issue by alebcay** [Homebrew/homebrew-core#39680]
**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [x] are reporting a bug others will be able to reproduce and not asking a question or requesting software. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh. To get software added or changed in Homebrew please file a [Pull Request](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)
- [x] have a problem with `brew install` (or `upgrade`, `reinstall`) a single, Homebrew/homebrew-core formula (not cask) on macOS? If it's a general `brew` problem please file this issue at Homebrew/brew: https://github.com/Homebrew/brew/issues/new/choose. If it's a Linux problem please file this issue at https://github.com/Homebrew/linuxbrew-core/issues/new/choose. If it's a `brew cask` problem please file this issue at https://github.com/Homebrew/homebrew-cask/issues/new/choose. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [ ] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

To help us debug your issue please explain:
- What you were trying to do (and why)
- What happened (include command output)
- What you expected to happen
- Step-by-step reproduction instructions (by running `brew install` commands)

---

See the full `gist-logs` output here: https://gist.github.com/alebcay/34e020ac7928198aa98b4ffb18d550a4

This issue arises when trying to `brew install --build-from-source hyperkit`.

It seems that `hyperkit`'s install process uses `opam` 2, which has introduced a sandbox feature. However, Homebrew also uses a sandbox during its build process, so `opam` freaks out as it cannot open a sandbox within a sandbox.

I have tried `opam init --disable-sandboxing`, but that did not fix anything.

This problem is holding back #39324.

**Comment by fxcoudert** [comment]
`ocaml`/`opam`-based formulas in Homebrew core are not in a very good state, because:

1. they tend to get broken / stop building when the ocaml compiler is updated (lack of stability in the compiler)
2. we don't really have someone in the team who understands it well (I think)

**Comment by human-bean** [comment]
For what it's worth, I was able to build hyperkit using the following, slightly modified, formula. Beware though, it is the result of trial and error.

```
class Hyperkit < Formula
  desc "Toolkit for embedding hypervisor capabilities in your application"
  homepage "https://github.com/moby/hyperkit"
  url "https://github.com/moby/hyperkit/archive/v0.20190201.tar.gz"
  sha256 "7ddd45f5ef20b6fa51141a2f7b98656ab1406c79cad449e2af05cb9218ba6c16"

  bottle do
    cellar :any_skip_relocation
    sha256 "575873fb7533d317e1044a2332d30c29eeef4d92653c2c16e780940e3695b32a" => :mojave
    sha256 "7a720e8737ac10ca41fb6db08e382eb40d87430a088891af694ce333e99d66ce" => :high_sierra
    sha256 "f4422d832468f23c3b845546205179edd61c4bb9a4c38ae39dd46018684f4a43" => :sierra
  end

  depends_on "aspcud" => :build
  depends_on "ocaml" => :build
  depends_on "opam" => :build
  depends_on :xcode => ["9.0", :build]

  depends_on "libev"

  resource "tinycorelinux" do
    url "https://dl.bintray.com/markeissler/homebrew/hyperkit-kernel/tinycorelinux_8.x.tar.gz"
    sha256 "560c1d2d3a0f12f9b1200eec57ca5c1d107cf4823d3880e09505fcd9cd39141a"
  end

  def install
    system "opam", "init", "--disable-sandboxing", "--no-setup"
    opam_dir = "#{buildpath}/.brew_home/.opam"
    ENV["CAML_LD_LIBRARY_PATH"] = "#{opam_dir}/system/lib/stublibs:#{Formula["ocaml"].opt_lib}/ocaml/stublibs"
    ENV["OPAMUTF8MSGS"] = "1"
    ENV["PERL5LIB"] = "#{opam_dir}/system/lib/perl5"
    ENV["OCAML_TOPLEVEL_PATH"] = "#{opam_dir}/system/lib/toplevel"
    ENV.prepend_path "PATH", "#{opam_dir}/system/bin"

    inreplace "#{buildpath}/repo/darwin/compilers/4.07.0.comp",
      '["./configure"', '["./configure" "-no-graph"' # Avoid X11

    ENV.deparallelize { system "opam", "switch", "create", "4.05.0" }

    system "opam", "config", "exec", "--",
           "opam", "install", "-y", "uri", "qcow.0.10.3", "conduit.1.0.0", "lwt.3.1.0",
           "qcow-tool", "mirage-block-unix.2.9.0", "conf-libev", "logs", "fmt",
           "mirage-unix", "prometheus-app", "cstruct-lwt.3.2.1"

    args = []
    args << "GIT_VERSION=#{version}"
    system "opam", "config", "exec", "--", "make", *args

    bin.install "build/hyperkit"
    man1.install "hyperkit.1"
  end

  test do
    assert_match(version.to_s, shell_output("#{bin}/hyperkit -v 2>&1"))

    if Hardware::CPU.features.include? :vmx
      resource("tinycorelinux").stage do |context|
        tmpdir = context.staging.tmpdir
        path_resource_versioned = Dir.glob(tmpdir.join("tinycorelinux_[0-9]*"))[0]
        cp(File.join(path_resource_versioned, "vmlinuz"), testpath)
        cp(File.join(path_resource_versioned, "initrd.gz"), testpath)
      end

      (testpath / "test_hyperkit.exp").write <<-EOS
        #!/usr/bin/env expect -d
        set KERNEL "./vmlinuz"
        set KERNEL_INITRD "./initrd.gz"
        set KERNEL_CMDLINE "earlyprintk=serial console=ttyS0"
        set MEM {512M}
        set PCI_DEV1 {0:0,hostbridge}
        set PCI_DEV2 {31,lpc}
        set LPC_DEV {com1,stdio}
        set ACPI {-A}
        spawn #{bin}/hyperkit $ACPI -m $MEM -s $PCI_DEV1 -s $PCI_DEV2 -l $LPC_DEV -f kexec,$KERNEL,$KERNEL_INITRD,$KERNEL_CMDLINE
        set pid [exp_pid]
        set timeout 20
        expect {
          timeout { puts "FAIL boot"; exec kill -9 $pid; exit 1 }
          "\\r\\ntc@box:~$ "
        }
        send "sudo halt\\r\\n";
        expect {
          timeout { puts "FAIL shutdown"; exec kill -9 $pid; exit 1 }
          "reboot: System halted"
        }
        expect eof
        puts "\\nPASS"
      EOS
      system "expect", "test_hyperkit.exp"
    end
  end
end
```

**Comment by alebcay** [comment]
Thanks for the insights. A slightly modified version of the slightly modified formula given above is now included in #39324. Had to do something similar for `corectl` since `ocaml` 4.04.2 is no longer available.

Also found that `-no-graph` is no longer a viable option for `ocaml` compiler config, so we have to `depends_on :x11 => :build` now. If anyone finds a way to remove the things that depend on x11 in the compiler (because we don't need them), let me know.

**Comment by alebcay** [comment]
Fixed in #41057 (https://github.com/Homebrew/homebrew-core/commit/9320c0d2316bbf5e7d9d1e752a76979949083136).
